core/embed/io/ble/stm32/ble.c

```diff
@@ -0,0 +1,505 @@
+                                                 uint32_t len) {
+  ble_driver_t *drv = &g_ble_driver;
+  if (!drv->initialized) {
+    return;
+  }
+
+  ble_event_t event = {.type = BLE_PAIRING_CANCELLED, .data_len = 0};
+  tsqueue_enqueue(&drv->event_queue, (uint8_t *)&event, sizeof(event), NULL);
+}
+
+static void ble_process_rx_msg(const uint8_t *data, uint32_t len) {
+  switch (data[0]) {
+    case INTERNAL_EVENT_STATUS:
+      ble_process_rx_msg_status(data, len);
+      break;
+    case INTERNAL_EVENT_PAIRING_REQUEST:
+      ble_process_rx_msg_pairing_request(data, len);
+      break;
+    case INTERNAL_EVENT_PAIRING_CANCELLED:
+      ble_process_rx_msg_pairing_cancelled(data, len);
+      break;
+    default:
+      break;
+  }
+}
+
+static void ble_process_data(const uint8_t *data, uint32_t len) {
+  ble_driver_t *drv = &g_ble_driver;
+
+  if (!drv->initialized) {
+    return;
+  }
+
+  if (len != BLE_RX_PACKET_SIZE) {
+    return;
+  }
+
+  tsqueue_enqueue(&drv->data_queue, data, len, NULL);
+}
+
+// background loop, called from systimer every 10ms
+static void ble_loop(void *context) {
+  (void)context;
+  ble_driver_t *drv = &g_ble_driver;
+
+  if (!drv->initialized) {
+    return;
+  }
+
+  if (nrf_is_running()) {
+    if (drv->ping_cntr == 0) {
+      ble_send_state_request();
+    }
+
+    drv->ping_cntr++;
+    if (drv->ping_cntr >= PING_PERIOD / LOOP_PERIOD_MS) {
+      drv->ping_cntr = 0;
+    }
+
+    uint8_t data[NRF_MAX_TX_DATA_SIZE] = {0};
+    if (tsqueue_dequeue(&drv->send_queue, data, NRF_MAX_TX_DATA_SIZE, NULL,
+                        NULL)) {
+      if (!nrf_send_msg(NRF_SERVICE_BLE, data, NRF_MAX_TX_DATA_SIZE, NULL,
+                        NULL)) {
+        tsqueue_enqueue(&drv->send_queue, data, NRF_MAX_TX_DATA_SIZE, NULL);
+      }
+    }
+
+    if (drv->mode_current != drv->mode_requested) {
+      if (drv->mode_requested == BLE_MODE_OFF) {
+        ble_send_advertising_off();
+        // if (drv->connected) {
+        //   nrf_send_disconnect();
+        // }
+      } else if (drv->mode_requested == BLE_MODE_CONNECTABLE) {
+        ble_send_advertising_on(true);
+      } else if (drv->mode_requested == BLE_MODE_PAIRING) {
+        ble_send_advertising_on(false);
+      }
+    }
+  } else {
+    drv->status_valid = false;
+  }
+}
+
+void ble_init(void) {
+  ble_driver_t *drv = &g_ble_driver;
+
+  if (drv->initialized) {
+    return;
+  }
+
+  memset(drv, 0, sizeof(ble_driver_t));
+
+  tsqueue_init(&drv->event_queue, drv->event_queue_entries,
+               (uint8_t *)drv->event_buffers, sizeof(ble_event_t),
+               EVENT_QUEUE_LEN);
+
+  tsqueue_init(&drv->data_queue, drv->data_queue_entries,
+               (uint8_t *)drv->data_buffers, BLE_RX_PACKET_SIZE,
+               DATA_QUEUE_LEN);
+
+  tsqueue_init(&drv->send_queue, drv->send_queue_entries,
+               (uint8_t *)drv->send_buffer, NRF_MAX_TX_DATA_SIZE,
+               SEND_QUEUE_LEN);
+
+  systimer_t *timer = systimer_create(ble_loop, NULL);
```

Comment:
Would disabling the timer in `ble_deinit` (or `ble_stop`) help with energy consumption or is that premature optimization? Currently neither function is ever called.

Reply:
not sure if it affects energy consumption, but it should be stopped. https://github.com/trezor/trezor-firmware/commit/191fa1a026fb38d579b593ff0ccff0fe2f32bccd

the deinits will eventually be called when puting the device in low power mode, or when jumping between bootloader and firmware. 

but currently we don't have ble in bootloader and low power mode is waiting for https://github.com/trezor/trezor-firmware/pull/4454